### ISSUE TITLE
multibody SDF parser does not properly parse packages outside of drake paths

### ISSUE THREAD

**Issue by edrumwri** [RobotLocomotion/drake#10099]
I'd like to use manipulation/util/geometry_inspector to manipulate a SDF file. The file is a part of an SDF "package" (attached), meaning that it has a package.xml in the directory root and resources located under the directory root. Drake fails to find the necessary resources _unless the package is put under the Drake path_, meaning that bazel needs to know all about it. This doesn't seem like a reasonable workflow for manipulating models not included with Drake.

Diving down, the reason that this happens is that LoadSdf() in sdf_parser.cc uses PackageMap::PopulateUpstreamToDrake() to locate the package root *which is expected to be located under the Drake root*. When the SDF is **not** located under the Drake root, the package name and path are not added to the PackageMap, and the SDF parser is then unable to find resources referenced from within the SDF file but located within the package.

Assigning to Sammy based on a nearby TODO in LoadSdf(). Feel free to reassign.

[toy_example_moon_v2.zip](https://github.com/RobotLocomotion/drake/files/2622432/toy_example_moon_v2.zip)

**Comment by jwnimmer-tri** [comment]
The right solution here is likely to have `drake::multibody::parsing::Parser` class expose the `PackageMap` object that will be used, so that the user may call one of the `PackageMap` methods that adds packages (before loading the SDF).  Doing that is likely quick surgical PR for anyone who wants this sooner rather than later.

**Comment by edrumwri** [comment]
You're advocating for this as both the "right solution" and the "quick surgical PR for anyone who wants this sooner rather than later". But the right solution to me seemed to be crawling up the directory tree, failing if no package.xml was found by the time root was reached. Why not that?

**Comment by sammy-tri** [comment]
I think I agree with Jeremy...  for loading packages which are part of drake, nothing would change.  For out of tree packages the code would look something like:
```
string model_package = "/path/to/package/";
string model_file = model_package + "urdf/my_model.urdf";
multibody::parsing::Parser parser(plant, scene_graph);
parser.package_map()->PopulateFromFolder(model_package);
parser.AddModelFromFile(model_file);
```

It's a little clunky but it would be simple to implement and have unambiguous behavior.  The alternative is crawling up the directory tree even for packages not in drake to search for `package.xml` files, and I'm not sure how I feel about that.

**Comment by jwnimmer-tri** [comment]
The question is how far up to go.

Crawling up from some directory into the forevermore of `/` reaches well beyond the files that a typical user should expect to have the tool sniff out.  I don't want `$HOME/package.xml` to be opened during this hunt just because I'm trying to load `$HOME/code/workspaces/project1/mybranch/foo/src/model/robot.sdf`.   For all we know, `$HOME/package.xml` could be entirely unrelated to ROS, and would throw a parser error.

We search up to Drake (when we can) because we know that the files we would find there are all part of the same inter-related source tree.

The correct general way is for the user to call `PackageMap::PopulateFromFolder("$HOME/code/workspaces/project1/mybranch")` or whatever, so that they can properly scope the extent of the search.

It's possible we'd also want sugar of "Open this SDF, using the package.xml from its immediate parent directory".  That would be fine, too.

**Comment by edrumwri** [comment]
Ok, searching for it could conceivably locate the wrong file. But I don't see how searching up to the Drake path doesn't suffer from that problem too (yes, we have control over the files, so the problem is less likely, but still possible) . It would seem the most robust solution would be to allow the user to specify the exact package path.

**Comment by jwnimmer-tri** [comment]
> ... I don't see how searching up to the Drake path doesn't suffer from that problem too ...

Searching within Drake by default is a convenience for our developers and users (it searches the installed Drake by default, iff you're loading from an installed SDF).  It produces a deterministic outcome, given some revision of the source tree, so isn't very dangerous.  The search is bounded and we know approximately what we're going to see (modulo PRs that add new package.xml files -- which are very rare).  That said, yes it is only a convenience default and perhaps being explicit all the time instead would be better.  You can carry that torch and work on removing it, if you like; I won't object.  However, I am comfortable with the current default.

> It would seem the most robust solution would be to allow the user to specify the exact package path.

Yup.  I think we're all in agreement that the mutable `PackageMap` should be exposed from `multibody::parsing::Parser`, and doing so would offer at least one satisfactory resolution to this issue.

**Comment by edrumwri** [comment]
Sounds good. Thanks!

**Comment by jwnimmer-tri** [comment]
To allow the user to update the PackageMap used during parsing, my specific proposal for the calling code would look like this:
```
Parser parser(&plant);
parser.package_map().Add("foo", "/path/to/foo");
parser.AddAllModelsFromFile("/path/to/foo/my.sdf");
```

**Comment by edrumwri** [comment]
Addressed by #10156.
